File: backend/ai_service/app/ai/voice/voices.py
"""Allowlist of selectable Cartesia TTS voices for the session voice loop.

The client (settings dropdown) sends a ``voice_id`` in the voice-loop ``start``
frame. That value is untrusted, so it is validated HERE before being handed to
``tts.stream_audio(voice_id=...)`` — an id not on this list falls back to the
configured default voice (``settings.cartesia_voice_id``) rather than reaching
Cartesia unchecked.

Mirror of the frontend option list (``frontend/src/constants/voices.ts``); keep
the two in sync. The names/genders are UI-only — only the ids matter here.
"""

from __future__ import annotations

from app.core.config import settings

# id -> human label (label unused server-side, kept for logging/debuggability).
SELECTABLE_VOICES: dict[str, str] = {
    "a5136bf9-224c-4d76-b823-52bd5efcffcc": "Jameson (Male)",
    "4f7f1324-1853-48a6-b294-4e78e8036a83": "Casper (Male)",
    "f039066f-cdb7-45ed-b51d-1034ae2f04a0": "Cindy Baker (Female)",
    "f9836c6e-a0bd-460e-9d3c-f7299fa60f94": "Caroline (Female)",
}


def resolve_voice_id(requested: str | None) -> str:
    """Return a safe Cartesia voice id for a client-requested value.

    A value on the allowlist is used as-is; anything else (None, empty, unknown,
    or tampered) resolves to the server default so a bad id can never fail a turn
    mid-stream as ``voice_not_found``.
    """
    if requested and requested in SELECTABLE_VOICES:
        return requested
    return settings.cartesia_voice_id

File: frontend/src/constants/voices.ts
// Selectable TTS voices for the hands-free session agent loop.
//
// These are Cartesia voice IDs — the value flows browser → gateway → ai_service
// and is handed to `stream_audio(voice_id=...)` (backend/ai_service/app/ai/voice/
// tts.py). The SAME id list is mirrored server-side as an allowlist in
// `app/ai/voice/voices.py`; keep the two in sync (an id absent from the server
// allowlist silently falls back to the server default voice).
//
// The choice is a rendering preference (which voice the agent SPEAKS with), not
// dining data, so it is persisted client-side (see stores/voicePrefStore.ts) and
// deliberately kept out of the dietary `Profile` / recommendation pipeline.

export interface VoiceOption {
  id: string
  name: string
  gender: 'Male' | 'Female'
}

export const VOICE_OPTIONS: VoiceOption[] = [
  { id: 'a5136bf9-224c-4d76-b823-52bd5efcffcc', name: 'Jameson', gender: 'Male' },
  { id: '4f7f1324-1853-48a6-b294-4e78e8036a83', name: 'Casper', gender: 'Male' },
  { id: 'f039066f-cdb7-45ed-b51d-1034ae2f04a0', name: 'Cindy Baker', gender: 'Female' },
  { id: 'f9836c6e-a0bd-460e-9d3c-f7299fa60f94', name: 'Caroline', gender: 'Female' },
]

// First option is the default when nothing is stored yet.
export const DEFAULT_VOICE_ID = VOICE_OPTIONS[0].id

// Guard a stored/loaded id against the current option set — a value no longer on
// the list (removed voice, tampered storage) resolves to the default.
export function isKnownVoiceId(id: string | null | undefined): id is string {
  return !!id && VOICE_OPTIONS.some((v) => v.id === id)
}
